app/plugin_health_check_test.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"testing"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHealthCheckJob(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	tearDown, _, _ := SetAppEnvironmentWithPlugins(t, []string{
+		`
+			package main
+
+			import (
+				"github.com/mattermost/mattermost-server/v5/model"
+				"github.com/mattermost/mattermost-server/v5/plugin"
+			)
+
+			type MyPlugin struct {
+				plugin.MattermostPlugin
+			}
+
+			func (p *MyPlugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*model.Post, string) {
+				panic("Uncaught error")
+			}
+
+			func main() {
+				plugin.ClientMain(&MyPlugin{})
+			}
+			`,
+	}, th.App, th.App.NewPluginAPI)
+	defer tearDown()
+
+	env := th.App.GetPluginsEnvironment()
+	job := env.GetPluginHealthCheckJob()
+	require.NotNil(t, job)
+	bundles := env.Active()
+	require.Equal(t, 1, len(bundles))
+
+	id := bundles[0].Manifest.Id
+
+	// First health check
+	hooks, err := env.HooksForPlugin(id)
+	require.Nil(t, err)
+	hooks.MessageWillBePosted(&plugin.Context{}, &model.Post{})
+	job.CheckPlugin(id)
+	bundles = env.Active()
+	require.Equal(t, 1, len(bundles))
+	require.Equal(t, id, bundles[0].Manifest.Id)
+	require.Equal(t, model.PluginStateRunning, env.GetPluginState(id))
+
+	// Second health check
+	hooks, err = env.HooksForPlugin(id)
+	require.Nil(t, err)
+	hooks.MessageWillBePosted(&plugin.Context{}, &model.Post{})
+	job.CheckPlugin(id)
+	bundles = env.Active()
+	require.Equal(t, 1, len(bundles))
+	require.Equal(t, id, bundles[0].Manifest.Id)
+	require.Equal(t, model.PluginStateRunning, env.GetPluginState(id))
+
+	// Third health check, plugin should be deactivated by the job
+	hooks, err = env.HooksForPlugin(id)
+	require.Nil(t, err)
+	hooks.MessageWillBePosted(&plugin.Context{}, &model.Post{})
```

Comment:
i believe these are not needed since the first call to `MessageWillBePosted` would've crashed the plugin, but no harm in keeping it :D

Reply:
We need to call it every time we want the plugin to crash though. The plugin gets restarted by the health check job each time we call `CheckPlugin`, so the hook does need to be called for each check here.

Reply:
ahh forgot about the restart 🤦‍♂. you didn't need it before because `Activate()` was causing the crash, cool makes sense 👌